Backup prunes transaction logs

phughk opened the pull request:
After a successful backup, the client will rotate logs and prune up to the last 2 log files.

phughk committed:
Backup prunes transaction logs

phughk commented:
This is the same issue as https://github.com/neo4j/neo4j/pull/11068

andrewkerr9000 commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/ForceCheckpointBackupService.java:
```diff
@@ -0,0 +1,132 @@
+        configParams.put( OnlineBackupSettings.online_backup_enabled.name(), Settings.FALSE );
+        configParams.put( GraphDatabaseSettings.keep_logical_logs.name(), Settings.FALSE );
+        GraphDatabaseFactory factory = ExternallyManagedPageCache.graphDatabaseFactoryWithPageCache( pageCache );
+        graphDatabaseAPI = (GraphDatabaseAPI) factory.newEmbeddedDatabaseBuilder( backupDirectory ).setConfig( configParams ).newGraphDatabase();
```
Why is a cast needed?

phughk replied:
Because the factory doesn't have that return type and a GraphDatabaseAPI is required to get access to the dependencies :man_shrugging: It happens like that in other areas of the code as well.

andrewkerr9000 replied:
It return a GraphDatabaseService. GraphDatabaseAPI is a subtype of GraphDatabaseService.

Probably can be improved with a strategically placed type parameter.

I'm making an addition to our refactoring column.

andrewkerr9000 commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/BackupStrategyCoordinator.java:
```diff
@@ -97,12 +103,27 @@ public void performBackup( OnlineBackupContext onlineBackupContext ) throws Comm
+        forceCheckpointBackupService.forceCheckpoint();
+        forceCheckpointBackupService.forcePrune( );
+        lifeSupport.stop();
+        lifeSupport.shutdown();
```
Do we really need a LifeSupport given that there's only one dependency?

Might be worth extracting the logic of init -> start -> (arbitrary Runnable) -> stop -> shutdown. Could be an instance method on LifeSupport, or a static method on a helper class that takes Collection\<Lifecycle\> and Runnable.

tinwelint replied:
You don't need to call init + start on LifeSupport, it does that internally. You should also have a shutdown in a finally, like so:

```
LifeSupport lifeSupport = new LifeSupport();
lifeSupport.add( forceCheckpointBackupService );
lifeSupport.start();
try
{
    forceCheckpointBackupService.forceRotation();
    forceCheckpointBackupService.forceCheckpoint();
    forceCheckpointBackupService.forcePrune();
}
finally
{
    lifeSupport.shutdown();
}
```

phughk replied:
To clarify the above, great comments, thank you. We have decided to not force checkpoints/pruning manually and instead use configs. Therefore the above code doesn't exist anymore

andrewkerr9000 commented on enterprise/backup/src/test/java/org/neo4j/backup/impl/BackupStrategyCoordinatorTest.java:
```diff
@@ -238,6 +244,44 @@ public void havingNoStrategiesCausesAllSolutionsFailedException() throws Command
+        // then
+        verify( forceCheckpointBackupService, never() ).forceCheckpoint();
+        verify( forceCheckpointBackupService, never() ).forcePrune();
+        verify( forceCheckpointBackupService, never() ).forceRotation();
```
Could `Mockito.verifyZeroInteractions( forceCheckpointBackupService )` replace these verifications?

phughk replied:
Removed

andrewkerr9000 commented on enterprise/backup/src/test/java/org/neo4j/backup/impl/BackupStrategyCoordinatorTest.java:
```diff
@@ -238,6 +244,44 @@ public void havingNoStrategiesCausesAllSolutionsFailedException() throws Command
+        // then
+        verify( forceCheckpointBackupService, times( 1 ) ).forceCheckpoint();
+        verify( forceCheckpointBackupService, times( 1 ) ).forcePrune();
+        verify( forceCheckpointBackupService, times( 1 ) ).forceRotation();
```
likewise

phughk committed:
Changed checkpoint to removing tx log files

tinwelint commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/ForceCheckpointBackupService.java:
```diff
@@ -0,0 +1,132 @@
+
+    private GraphDatabaseAPI graphDatabaseAPI;
+
+    ForceCheckpointBackupService( PageCache pageCache, FileSystemAbstraction fileSystemAbstraction, Config config, File backupDirectory )
```
I believe backup already starts a database internally to do the recovery of concurrent transactions while streaming the store files, after the streaming. Can't that db just do the pruning? Seems unnecessary to open the db twice.

tinwelint commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/ForceCheckpointBackupService.java:
```diff
@@ -0,0 +1,132 @@
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.kernel.lifecycle.Lifecycle;
+
+class ForceCheckpointBackupService implements Lifecycle
```
Couldn't this be an `AutoCloseable` instead? Does it need to be a `Lifecycle` specifically somehow?

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -617,6 +619,7 @@ private NeoStoreTransactionLogModule buildTransactionLogs( LogFiles logFiles, Co
         final CheckPointerImpl checkPointer = new CheckPointerImpl(
                 transactionIdStore, threshold, storageEngine, logPruning, appender, databaseHealth, logProvider,
                 tracers.checkPointTracer, ioLimiter, storeCopyCheckPointMutex );
+        dependencies.satisfyDependencies( checkPointer );
```
This is already made available [here](https://github.com/neo4j/neo4j/blob/3.4/community/kernel/src/main/java/org/neo4j/kernel/NeoStoreTransactionLogModule.java#L70).

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -601,9 +601,11 @@ private NeoStoreTransactionLogModule buildTransactionLogs( LogFiles logFiles, Co
 
         final LogRotation logRotation =
                 new LogRotationImpl( monitors.newMonitor( LogRotation.Monitor.class ), logFiles, databaseHealth );
+        dependencies.satisfyDependencies( logRotation );
```
This is already made available [here](https://github.com/neo4j/neo4j/blob/3.4/community/kernel/src/main/java/org/neo4j/kernel/NeoStoreTransactionLogModule.java#L70).

tinwelint commented on enterprise/backup/pom.xml:
```diff
@@ -81,6 +81,11 @@
     </dependency>
+    <dependency>
+      <groupId>org.neo4j</groupId>
+      <artifactId>neo4j-tools</artifactId>
```
This dependency is super backwards neo4j-tools is on top of the component-chain... please break it in some way. Why was it introduced? Is it only for the test utility BackupTransactionLogFilesHelper?

tinwelint commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/ForceCheckpointBackupService.java:
```diff
@@ -0,0 +1,132 @@
+        logPruning.pruneLogs( logFiles.getHighestLogVersion() );
+    }
+
+    void forceRotation()
```
Why is log rotation forced? Check point does a log rotation and what's the issue with keeping 1 transaction log around?

tinwelint commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/ForceCheckpointBackupService.java:
```diff
@@ -0,0 +1,132 @@
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
```
Isn't the proper way of letting this method throw IOException and eventually rethrow as CommandFailed further up? Observing this catch-rethrow-as-RuntimeException It doesn't look like you've thought about failure scenarios.

tinwelint commented on enterprise/backup/src/main/java/org/neo4j/backup/impl/ForceCheckpointBackupService.java:
```diff
@@ -0,0 +1,132 @@
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
```
Isn't the proper way of letting this method throw IOException and eventually rethrow as CommandFailed further up? Observing this catch-rethrow-as-RuntimeException It doesn't look like you've thought about failure scenarios.

tinwelint commented on enterprise/backup/src/test/java/org/neo4j/backup/impl/BackupTransactionLogFilesHelper.java:
```diff
@@ -0,0 +1,87 @@
+        return LogFilesBuilder.activeFilesBuilder( backupDir, fileSystemAbstraction, pageCache ).build();
+    }
+
+    Pair<List<LogEntry[]>,List<CheckPoint>> logEntriesAndCheckpoints( File logFile )
```
I believe to break the neo4j-tools dependency, this method could be replaced by:

```
    Pair<List<LogEntry[]>,List<CheckPoint>> logEntriesAndCheckpoints( FileSystemAbstraction fs, File logFile ) throws IOException
    {
        List<LogEntry[]> transactions = new ArrayList<>();
        List<CheckPoint> checkPoints = new ArrayList<>();

        StoreChannel logFileChannel = fs.open( logFile, OpenMode.READ );
        LogHeader logHeader = readLogHeader( ByteBuffer.allocate( LOG_HEADER_SIZE ), logFileChannel, true, logFile );
        LogEntryReader<ReadableClosablePositionAwareChannel> entryReader = new VersionAwareLogEntryReader<>( new RecordStorageCommandReaderFactory(), STRICT );
        try ( ReadAheadLogChannel channel = new ReadAheadLogChannel(
                new PhysicalLogVersionedStoreChannel( logFileChannel, logHeader.logVersion, logHeader.logFormatVersion ), NO_MORE_CHANNELS );
                LogEntryCursor cursor = new LogEntryCursor( entryReader, channel ) )
        {
            List<LogEntry> tx = new ArrayList<>();
            while ( cursor.next() )
            {
                LogEntry entry = cursor.get();
                if ( entry.getType() == LogEntryByteCodes.CHECK_POINT )
                {
                    checkPoints.add( entry.as() );
                }
                else
                {
                    tx.add( entry );
                    if ( entry.getType() == LogEntryByteCodes.TX_COMMIT )
                    {
                        transactions.add( tx.toArray( new LogEntry[tx.size()] ) );
                    }
                }
            }
        }
        return Pair.of( transactions, checkPoints );
    }
```

tinwelint replied:
Alternatively the TransactionLogAnalyzer could've moved to kernel

tinwelint commented on enterprise/backup/src/test/java/org/neo4j/backup/impl/BackupTransactionLogFilesHelper.java:
```diff
@@ -0,0 +1,87 @@
+                listOfCheckpoints.add( checkpoint );
+            }
+        };
+        InvalidLogEntryHandler invalidLogEntryHandler = new InvalidLogEntryHandler()
```
Also take it as a general note to check prior art before implementing something new. The less code we have, the better we're off, right :) E.g. `InvalidLogEntryHandler.STRICT` could might as well have been used.

Also please read javadoc, they are there to help us all understand each others code and design, because the `handleInvalidEntry` wasn't even supposed to throw exception.

tinwelint reviewed (changes requested):
It looks like this change was rushed and that you didn't take your time to fully understand all changes you've made. Also for simplicity I'd think that instead a change like:

```
diff --git a/enterprise/backup/src/main/java/org/neo4j/backup/impl/BackupProtocolService.java b/enterprise/backup/src/main/java/org/neo4j/backup/impl/BackupProtocolService.java
index a9f1e8c..cbdb073 100644
--- a/enterprise/backup/src/main/java/org/neo4j/backup/impl/BackupProtocolService.java
+++ b/enterprise/backup/src/main/java/org/neo4j/backup/impl/BackupProtocolService.java
@@ -156,7 +156,7 @@ public class BackupProtocolService
             long lastCommittedTx = -1;
             StoreCopyClient storeCopier = new StoreCopyClient( targetDirectory.toFile(), tuningConfiguration,
                     loadKernelExtensions(), logProvider, fileSystem, pageCache,
-                    monitors.newMonitor( StoreCopyClientMonitor.class, getClass() ), forensics );
+                    monitors.newMonitor( StoreCopyClientMonitor.class, getClass() ), forensics, true );
             FullBackupStoreCopyRequester storeCopyRequester =
                     new FullBackupStoreCopyRequester( sourceHostNameOrIp, sourcePort, timeout, forensics, monitors );
             storeCopier.copyStore(
diff --git a/enterprise/com/src/main/java/org/neo4j/com/storecopy/StoreCopyClient.java b/enterprise/com/src/main/java/org/neo4j/com/storecopy/StoreCopyClient.java
index 2a55093..5250298 100644
--- a/enterprise/com/src/main/java/org/neo4j/com/storecopy/StoreCopyClient.java
+++ b/enterprise/com/src/main/java/org/neo4j/com/storecopy/StoreCopyClient.java
@@ -31,6 +31,7 @@ import java.util.stream.Stream;
 
 import org.neo4j.com.Response;
 import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.factory.GraphDatabaseBuilder;
 import org.neo4j.graphdb.factory.GraphDatabaseSettings;
 import org.neo4j.helpers.CancellationRequest;
 import org.neo4j.helpers.collection.Visitor;
@@ -89,17 +90,19 @@ public class StoreCopyClient
     private final PageCache pageCache;
     private final StoreCopyClientMonitor monitor;
     private final boolean forensics;
+    private final boolean pruneLogs;
     private final FileMoveProvider fileMoveProvider;
 
     public StoreCopyClient( File storeDir, Config config, Iterable<KernelExtensionFactory<?>> kernelExtensions, LogProvider logProvider,
-            FileSystemAbstraction fs, PageCache pageCache, StoreCopyClientMonitor monitor, boolean forensics )
+            FileSystemAbstraction fs, PageCache pageCache, StoreCopyClientMonitor monitor, boolean forensics, boolean pruneLogs )
     {
-        this( storeDir, config, kernelExtensions, logProvider, fs, pageCache, monitor, forensics, new FileMoveProvider( pageCache,
+        this( storeDir, config, kernelExtensions, logProvider, fs, pageCache, monitor, forensics, pruneLogs, new FileMoveProvider( pageCache,
                 fs ) );
     }
 
-    public StoreCopyClient( File storeDir, Config config, Iterable<KernelExtensionFactory<?>> kernelExtensions, LogProvider logProvider,
-            FileSystemAbstraction fs, PageCache pageCache, StoreCopyClientMonitor monitor, boolean forensics, FileMoveProvider fileMoveProvider )
+    StoreCopyClient( File storeDir, Config config, Iterable<KernelExtensionFactory<?>> kernelExtensions, LogProvider logProvider,
+            FileSystemAbstraction fs, PageCache pageCache, StoreCopyClientMonitor monitor, boolean forensics, boolean pruneLogs,
+            FileMoveProvider fileMoveProvider )
     {
         this.storeDir = storeDir;
         this.config = config;
@@ -109,6 +112,7 @@ public class StoreCopyClient
         this.pageCache = pageCache;
         this.monitor = monitor;
         this.forensics = forensics;
+        this.pruneLogs = pruneLogs;
         this.fileMoveProvider = fileMoveProvider;
     }
 
@@ -258,17 +262,22 @@ public class StoreCopyClient
     {
         ExternallyManagedPageCache.GraphDatabaseFactoryWithPageCacheFactory factory =
                 ExternallyManagedPageCache.graphDatabaseFactoryWithPageCache( pageCache );
-        return factory
+        GraphDatabaseBuilder builder = factory
                 .setKernelExtensions( kernelExtensions )
                 .setUserLogProvider( NullLogProvider.getInstance() )
-                .newEmbeddedDatabaseBuilder( tempStore.getAbsoluteFile() )
-                .setConfig( "dbms.backup.enabled", Settings.FALSE )
+                .newEmbeddedDatabaseBuilder( tempStore.getAbsoluteFile() ).setConfig( "dbms.backup.enabled", Settings.FALSE )
                 .setConfig( GraphDatabaseSettings.pagecache_warmup_enabled, Settings.FALSE )
                 .setConfig( GraphDatabaseSettings.logs_directory, tempStore.getAbsolutePath() )
                 .setConfig( GraphDatabaseSettings.keep_logical_logs, Settings.TRUE )
                 .setConfig( GraphDatabaseSettings.logical_logs_location, tempStore.getAbsolutePath() )
-                .setConfig( GraphDatabaseSettings.allow_upgrade, config.get( GraphDatabaseSettings.allow_upgrade ).toString() )
-                .newGraphDatabase();
+                .setConfig( GraphDatabaseSettings.allow_upgrade, config.get( GraphDatabaseSettings.allow_upgrade ).toString() );
+        if ( pruneLogs )
+        {
+            builder = builder
+                    .setConfig( GraphDatabaseSettings.keep_logical_logs, Settings.FALSE )
+                    .setConfig( GraphDatabaseSettings.logical_log_rotation_threshold, "1m" ); // <-- smallest possible threshold
+        }
+        return builder.newGraphDatabase();
     }
 
     private StoreWriter decorateWithProgressIndicator( final StoreWriter actual )
diff --git a/enterprise/com/src/test/java/org/neo4j/com/storecopy/StoreCopyClientTest.java b/enterprise/com/src/test/java/org/neo4j/com/storecopy/StoreCopyClientTest.java
index 870e34d..299f5e8 100644
--- a/enterprise/com/src/test/java/org/neo4j/com/storecopy/StoreCopyClientTest.java
+++ b/enterprise/com/src/test/java/org/neo4j/com/storecopy/StoreCopyClientTest.java
@@ -137,7 +137,7 @@ public class StoreCopyClientTest
         StoreCopyClient copier =
                 new StoreCopyClient( copyDir, Config.defaults(), loadKernelExtensions(), NullLogProvider.getInstance(),
                         fileSystem,
-                        pageCache, storeCopyMonitor, false );
+                        pageCache, storeCopyMonitor, false, false );
 
         final GraphDatabaseAPI original =
                 (GraphDatabaseAPI) startDatabase( originalDir );
@@ -184,7 +184,7 @@ public class StoreCopyClientTest
         Config config = Config.defaults( logical_logs_location, copyCustomLogFilesLocation.getName() );
         StoreCopyClient copier = new StoreCopyClient(
                 copyDir, config, loadKernelExtensions(), NullLogProvider.getInstance(), fileSystem, pageCache,
-                new StoreCopyClientMonitor.Adapter(), false );
+                new StoreCopyClientMonitor.Adapter(), false, false );
 
         GraphDatabaseAPI original = (GraphDatabaseAPI) new TestGraphDatabaseFactory()
                 .newEmbeddedDatabaseBuilder( originalDir )
@@ -249,7 +249,7 @@ public class StoreCopyClientTest
         PageCache pageCache = pageCacheRule.getPageCache( fileSystem );
         StoreCopyClient copier = new StoreCopyClient(
                 copyDir, Config.defaults(), loadKernelExtensions(), NullLogProvider.getInstance(), fileSystem, pageCache,
-                storeCopyMonitor, false );
+                storeCopyMonitor, false, false );
 
         final GraphDatabaseAPI original = (GraphDatabaseAPI) startDatabase( originalDir );
 
@@ -297,7 +297,7 @@ public class StoreCopyClientTest
 
         StoreCopyClient copier =
                 new StoreCopyClient( backupStore, Config.defaults(), loadKernelExtensions(), NullLogProvider
-                        .getInstance(), fileSystem, pageCache, new StoreCopyClientMonitor.Adapter(), false );
+                        .getInstance(), fileSystem, pageCache, new StoreCopyClientMonitor.Adapter(), false, false );
         CancellationRequest falseCancellationRequest = () -> false;
         StoreCopyClient.StoreCopyRequester storeCopyRequest =
                 requestFactory.create( (GraphDatabaseAPI) initialDatabase, initialStore, fileSystem, false );
@@ -325,7 +325,7 @@ public class StoreCopyClientTest
         GraphDatabaseService initialDatabase = createInitialDatabase( initialStore );
         StoreCopyClient copier =
                 new StoreCopyClient( backupStore, Config.defaults(), loadKernelExtensions(), NullLogProvider
-                        .getInstance(), fileSystem, pageCache, new StoreCopyClientMonitor.Adapter(), false );
+                        .getInstance(), fileSystem, pageCache, new StoreCopyClientMonitor.Adapter(), false, false );
         CancellationRequest falseCancellationRequest = () -> false;
 
         RuntimeException exception = new RuntimeException( "Boom!" );
@@ -362,7 +362,7 @@ public class StoreCopyClientTest
         Config config = Config.defaults( record_format, recordFormatsName );
         StoreCopyClient copier = new StoreCopyClient(
                 copyDir, config, loadKernelExtensions(), NullLogProvider.getInstance(), fileSystem, pageCache,
-                new StoreCopyClientMonitor.Adapter(), false );
+                new StoreCopyClientMonitor.Adapter(), false, false );
 
         final GraphDatabaseAPI original = (GraphDatabaseAPI) startDatabase( originalDir, recordFormatsName );
         StoreCopyClient.StoreCopyRequester storeCopyRequest = requestFactory.create( original, originalDir,
diff --git a/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveBranchThenCopy.java b/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveBranchThenCopy.java
index f0d209b..6cbfeda 100644
--- a/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveBranchThenCopy.java
+++ b/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveBranchThenCopy.java
@@ -96,7 +96,7 @@ public class SwitchToSlaveBranchThenCopy extends SwitchToSlave
                 masterClientResolver,
                 monitor,
                 new StoreCopyClient( storeDir, config, kernelExtensions, logService.getUserLogProvider(),
-                        fileSystemAbstraction, pageCache, storeCopyMonitor, false ),
+                        fileSystemAbstraction, pageCache, storeCopyMonitor, false, false ),
                 neoDataSourceSupplier,
                 transactionIdStoreSupplier,
                 slaveServerFactory,
diff --git a/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveCopyThenBranch.java b/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveCopyThenBranch.java
index 2baaaef..72bd0ce 100644
--- a/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveCopyThenBranch.java
+++ b/enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlaveCopyThenBranch.java
@@ -99,7 +99,7 @@ public class SwitchToSlaveCopyThenBranch extends SwitchToSlave
                 masterClientResolver,
                 monitor,
                 new StoreCopyClient( storeDir, config, kernelExtensions, logService.getUserLogProvider(),
-                        fileSystemAbstraction, pageCache, storeCopyMonitor, false ),
+                        fileSystemAbstraction, pageCache, storeCopyMonitor, false, false ),
                 neoDataSourceSupplier,
                 transactionIdStoreSupplier,
                 slaveServerFactory,
```

would accomplish what you seek, without the added overhead of the additional code, dependencies and extra db start/shutdown.

phughk committed:
Mattias' changes using modified config

tinwelint commented:
@phughk any progress on this?

phughk committed:
WIP

phughk committed:
WIP

Backup TransactionLogCatchupWriter is read only, so only 1 rotation
allowed - during shutdown

Why read only?

phughk commented:
Regarding the diff provided above, it won't work since the transactions are pulled in batches and the checks for rotation/pruning are done between batches. So under the diff you have provided the rotation should only work after shutdown.

phughk commented:
Moved
